### ISSUE TITLE
Code Review: Accessibility features

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/john-travoltage#128]
In preparation for deployment, accessibility need another code review.  A full code review shouldn't be necessary as accessibility code should be limited to the view.  I modified the normal code review checklist to cover things that should be checked.

*NOTE! Prior to doing a code review, copy this checklist to a GitHub issue for the repository being reviewed.*

PhET code-review checklist
=============

**Build and Run Checks**

- [ ] Does the sim build without warnings or errors?
- [ ] Does the sim start up? (requirejs and built versions)
- [ ] Does the sim experience any assertion failures? (run with query parameter 'ea')
- [ ] Does the sim pass a scenery fuzzer test? (run with query parameter 'fuzzMouse&ea')

**Internationalization**
- [ ] Are there any strings that are not being internationalized? (run with query parameter 'stringTest=x', you should see nothing but 'x' strings)
  - Accessibility strings should *not* be internationalized.  We want community feedback on descriptions before they are submitted for internationalization.  We also need additional support in the translation utility to let translators know what can be translated.
- [ ] Does the sim layout gracefully handle internationalized strings that are twice as long as the English strings? (run with query parameter 'stringTest=double')
- [ ] Does the sim layout gracefully handle internationalized strings that are exceptionally long? (run with query parameter 'stringTest=long')
- [ ] Does the sim layout gracefully handle internationalized strings that are shorter than the English strings? (run with query parameter 'stringTest=X')
- [ ] Does the sim stay on the sim page (doesn't redirect to an external page) when running with the query parameter 'stringTest=xss' (test passes if sim does not redirect, OK if sim crashes or fails to fully start). Only test on one desktop platform.
- [ ] Make sure the string keys are all perfect, they are difficult to change after 1.0.0 is published.  Strings keys should
generally match the values, such as `{binaryProbability: "Binary Probability"}`. Screen names should use `screen.screenName` instead of
camelcase.  Message patterns and long paragraphs will also use a different pattern.

**Repository structure**

- [ ] Are all required files and directories present?

      For a sim repository named “my-repo”, the general structure should look like this (where assets/, audio/ or images/ may be omitted if the sim doesn’t have those types of assets).

```js
   my-repo/
      assets/
      audio/
         license.json
      doc/
         model.md
         implementation-notes.md
      images/
         license.json
      js/
         (see section below)
      .gitignore
      my-repo_en.html
      my-repo-strings_en.json
      Gruntfile.js
      LICENSE
      package.json
      README.md
```

   For a common-code repository, the structure is similar, but some of the files and directories may not be present if the repo doesn’t have audio, images, strings, or a demo application.

- [ ] Is the js/ directory properly structured?

      All JavaScript source should be in the js/ directory. There should be a subdirectory for each screen (this also applies for single-screen sims, where the subdirectory matches the repo name).  For a multi-screen sim, code shared by 2 or more screens should be in a js/common/ subdirectory. Model and view code should be in model/ and view/ subdirectories for each screen and common/.  For example, for a sim with screens “Introduction” and “Lab”, the general directory structure should look like this:

```js
   my-repo/
      js/
         common/
            model/
            view/
         introduction/
            model/
            view/
         lab/
            model/
            view/
         my-repo-config.js
         my-repo-main.js
         myRepo.js
 ```

- [ ] Is there a file in assets/ for every resource file in audio/ and images/? Note that there is *not necessarily* a
1:1 correspondence between asset and resource files; for example, several images may be in the same .ai file.
- [ ] Was the README.md generated by `grunt published-README` or `grunt unpublished-README`?
- [ ] Does package.json refer to any dependencies that are not used by the sim?
- [ ] Does sim-config.js refer to any dependencies that are not used by the sim?
- [ ] Is the LICENSE file correct? (GPL v3 for sims, MIT for common code)
- [ ] Does .gitignore match the one in simula-rasa?
- [ ] Are the credits in sim-main.js correct?
- [ ] Are there git repository branches that are no longer used and should be pruned?

**Coding conventions**

- [ ] Does the code meet PhET's code-style guidelines, as described in the [“Coding Style Guidelines” section of "PhET Development Overview"](https://docs.google.com/document/d/1odXkliRagq0zuf1_NdOtQ2BrkC9hwlISnPi2y-dKdrk/edit#heading=h.1oxr3ptyo50w)?
- [ ] Is the code formatted according to PhET conventions? See [phet-idea-code-style.xml](https://github.com/phetsims/joist/blob/master/util/phet-idea-codestyle.xml) for IntelliJ IDEA code style.

**Documentation**

- [ ] Are documentation conventions followed, as described in the [“Coding Style Guidelines” section of the PhET Development Overview](https://docs.google.com/document/d/1odXkliRagq0zuf1_NdOtQ2BrkC9hwlISnPi2y-dKdrk/edit#heading=h.1oxr3ptyo50w)?
- [ ] Does model.md adequately describe the model, in terms appropriate for teachers?
- [ ] Does implementation-notes.md adequately describe the implementation, with an overview that will be useful to future maintainers?
- [ ] Are sim-specific query parameters (if any) identified and documented in one .js file in js/common/ or js/ (if there is no common/)? The .js file should be named `{{REPO}}QueryParameters`, for example ArithmeticQueryParameters for the aritmetic repository.

**Math Libraries**

- [ ] Check that `dot.Util.roundSymmetric` is used instead of `Math.round`. `Math.round` does not treat positive and negative numbers symmetrically, see https://github.com/phetsims/dot/issues/35#issuecomment-113587879.
- [ ] `dot.Util.toFixed` or `dot.Util.toFixedNumber` should be used instead of `toFixed`. JavaScript's `toFixed` is notoriously buggy. Behavior differs depending on browser, because the spec doesn't specify whether to round or floor.
- [ ] Check that random numbers are generated using `phet.joist.random`, and are doing so after modules are declared (non-statically).  For instance, the following methods (and perhaps others) should not be used:
* `Math.random`
* `_.shuffle`
* `_.sample`
* `_.random`
* `new Random()`

**Organization, Readability, Maintainability**

- [ ] Does the organization and structure of the code make sense? Do the model and view contain types that you would expect (or guess!) by looking at the sim? Do the names of things correspond to the names that you see in the user interface?
- [ ] Are appropriate design patterns used?
- [ ] Is inheritance used where appropriate? Does the type hierarchy make sense?
- [ ] Is there any unnecessary coupling? (e.g., by passing large objects to constructors, or exposing unnecessary properties/functions)
- [ ] Is there too much unnecessary decoupling? (e.g. by passing all of the properties of an object independently instead of passing the object itself)?
- [ ] Are the source files reasonable in size? Scrutinize large files with too many responsibilities - can responsibilities be broken into smaller delegates?
- [ ] Are any significant chunks of code duplicated? This will be checked manually as well as with https://github.com/danielstjules/jsinspect
- [ ] Is there anything that should be generalized and migrated to common code?
- [ ] Are there any TODO or FIXME comments in the code?  They should be addressed or promoted to GitHub issues.
- [ ] Does the implementation rely on any specific constant values that are likely to change in the future? Identify constants that might be changed in the future. (Use your judgement about which constants are likely candidates.)
Does changing the values of these constants break the sim? For example, see https://github.com/phetsims/plinko-probability/issues/84.
- [ ] The simulation should use Property instead of PropertySet
- [ ] Are all dependent properties modeled as DerivedProperty instead of Property?

**Performance, Usability**

- [ ] Does the sim perform as desired across the range of supported platforms? (eg, not too slow on slow platforms, not too fast on fast platforms)
- [ ] If the sim uses WebGL, does it have a fallback? Does the fallback perform reasonably well? (run with query parameter 'webgl=false')
- [ ] Are UI components sufficiently responsive? (especially continuous UI components, such as sliders)
- [ ] Are pointer areas optimized, especially for touch? (run with query parameter 'showPointerAreas')
- [ ] Do pointer areas overlap? (run with query parameter 'showPointerAreas')
- [ ] Is the timestep `dt` capped appropriately?  Try switching applications or browser tabs, then switch back.  Did the model take one big/long/awkward step forward?  If so, `dt` may need to be capped.  Example from `faradays-law.FaradaysLawModel`:

```js
// Cap large dt values, which can occur when the tab containing
// the sim had been hidden and then re-shown
dt = Math.min( 0.1, dt );
```

**Memory Leaks**

- [ ] Does a heap comparison using Chrome Developer Tools indicate a memory leak? (Describing this process is beyond the scope of this document.)
- [ ] For each common-code component (sun, scenery-phet, vegas, …) that opaquely registers observers or listeners, is there a call to that component’s dispose function, or documentation about why dispose is unnecessary?
- [ ] Are there leaks due to registering observers or listeners? These guidelines should be followed, or documentation added about why following them is not necessary:
	- [ ] AXON: `Property.link` is accompanied by `Property.unlink`.
	- [ ] AXON: `PropertySet.link` is accompanied by `PropertySet.unlink`.
	- [ ] AXON: Creation of `DerivedProperty` is accompanied by `dispose`.
	- [ ] AXON: Creation of `Multilink` is accompanied by `dispose`.
	- [ ] AXON: `Events.on` is accompanied by `Events.off`.
	- [ ] AXON: `Emitter.addListener` is accompanied by `Emitter.removeListener`.
	- [ ] SCENERY: `Node.on` is accompanied by `Node.off`
	- [ ] TANDEM: `tandem.addInstance` is accompanied by `tandem.removeInstance`.
- [ ] Do all types that require a `dispose` function have one? This should expose a public `dispose` function that calls
`this.disposeMyType()`, where `disposeMyType` is a private function declared in the constructor.  `MyType` should exactly
match the filename.

**PhET-iO**

- [ ] If the simulation is supposed to be instrumented for PhET-iO, please see [How to Instrument a PhET Simulation for PhET-iO]
(https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md)
for the PhET-iO development process.

**Comment by jessegreenberg** [comment]
I will take one more pass through the list and check the code before assigning to another.

**Comment by jessegreenberg** [comment]
@ariel-phet and @emily-phet this issue will be used to track the code review.  @ariel-phet can you please assign and prioritize the code review?

We could do a review in two ways:
 - A full code review - longer review bringing the sim up to standards of the latest code review checklist.
 - Only review code related to the accessibility features - shorter review, just touching accessibility blocks of code in the view.  If we go this route, we can reduce the above check list.

**Comment by emily-phet** [comment]
@ariel-phet - Last we spoke about this, you mentioned JO might be a good fit for a code review.

**Comment by ariel-phet** [comment]
@jessegreenberg @emily-phet will talk with JO, I know currently CL:B code review was priority for him due to AA departure.

**Comment by emily-phet** [comment]
@ariel-phet Thanks!

**Comment by ariel-phet** [comment]
@jonathanolson is going to go through the sim with an eye for review issues on a11y features, if he runs into code review issues for the code in general (PhET standards) he will just fix not create issues.

**Comment by jessegreenberg** [comment]
A new type called AccessibleNode was recently checked into scenery and this sim should use it for code related to accessibility.  It should reduce much of the boiler plate code that was previously required for setting accessible content.  Code review should be temporarily on hold until #132 is complete, which I will work on at high priority so this can continue as soon as possible.

This should not take much time, I estimate roughly an hour.

**Comment by jessegreenberg** [comment]
I will assign back to @jonathanolson when #132 is complete.

**Comment by jessegreenberg** [comment]
Alright, #132 is done, assigning back to @jonathanolson.

**Comment by jonathanolson** [comment]
@jessegreenberg, what common code should I review as part of this? I'd plan on scenery/js/accessibility/* and other Scenery code that interacts with it, but I didn't know if there were other types I should be looking at.

I think discussions about how to handle abstraction of status/alerts and common accessible components (like AccessibleSlider or button mixins) would be a good next step.

Additionally, having some centralized documentation for PhET developers on how browsers and accessibility tools interpret our DOM manipulations would be helpful, unless those DOM manipulations are encapsulated in common code.

**Comment by jonathanolson** [comment]
I've basically completed the sim-specific review for accessibility, although as issues are handled I'll want to re-review the modified code.

I haven't done a fully in-depth review for the simulation in general, particularly for code that didn't look accessibility-related.

**Comment by jessegreenberg** [comment]
>  what common code should I review as part of this? I'd plan on scenery/js/accessibility/* 

I don't think that is necessary for this issue.  scenery/js/accessibility/* contains a directory reader/*, which contains some prototype code that is not not going to be used in production code for quite some time, and is not used by john-travoltage.

The other types in scenery/js/accessibility/* are AccessibleNode, which is being worked on in https://github.com/phetsims/scenery/issues/596, AccessiblePeer, and AccessibleInstance, and I am not sure if those require review at this time.

**Comment by jonathanolson** [comment]
Sounds good, unassigning.

**Comment by jessegreenberg** [comment]
All remaining issues are either related to sonification or are are ready for further review.  There are 4 still open that involve a11y.

**Comment by jessegreenberg** [comment]
Closing, remaining issues involve further review.
